app/services/sso/apache.rb

```diff
@@ -30,6 +30,7 @@ def support_fallback?
     def authenticated?
       return false unless (self.user = request.env[CAS_USERNAME])
       attrs = { :login => user }.merge(additional_attributes)
+      attrs[:mail] = nil if attrs[:mail] == "(null)"
```

Comment:
null attributes are returned as "(null)", which was breaking the email on creation of external account without email. Now it works properly and also also added this to the relevant test.

Reply:
ftr, the `(null)` comes from apache's mod_headers[1].

[1] - https://github.com/apache/httpd/blob/fe4ade610c750d63c3bbfe0d07c5a37f2d5cb9f0/modules/metadata/mod_headers.c#L204

Reply:
The test you modified was passing before this change, so I'm not sure how you reached this conclusion. Could you provide more details?

Reply:
Yes, I had a testing instance hooked up to the ipa server (thanks @adamruzicka) which had a user without email. The instance had my patch applied but when trying to log in as the user without the email - the account was not yet created on the foreman server - the login attempt failed and the backtrace appeared in the logs. The backtrace was caused by error in the email validator, because it was receiving `(null)` in text form as the email value.

Reply:
Before I modified the test, the email was implicitly nil, so the email validation did not break and it failed to catch this scenario.

Reply:
@adamruzicka is correct that this is something Apache does. We already have some code there this happens:
https://github.com/theforeman/foreman/blob/faee8af3f8f2e4d41cb6f1ba39c20ec72b17b0d8/app/services/foreman/client_certificate.rb#L11